util/patch/patch.go

```diff
@@ -86,49 +95,63 @@ func (h *Helper) Patch(ctx context.Context, resource runtime.Object) error {
 		return errors.Errorf("expected non-nil resource")
 	}
 
-	// If the object is already unstructured, we need to perform a deepcopy first
-	// because the `DefaultUnstructuredConverter.ToUnstructured` function returns
-	// the underlying unstructured object map without making a copy.
-	if _, ok := resource.(runtime.Unstructured); ok {
-		resource = resource.DeepCopyObject()
-	}
-
 	// Convert the resource to unstructured to compare against our before copy.
-	after, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resource)
+	raw, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resource.DeepCopyObject())
 	if err != nil {
 		return err
 	}
+	unstructuredResource := &unstructured.Unstructured{Object: raw}
+
+	// Determine if the object has status.
+	afterHasStatus := unstructuredHasStatus(unstructuredResource)
+	if afterHasStatus {
+		if h.opts.IncludeStatusObservedGeneration {
+			// Set status.observedGeneration if we're asked to do so.
+			if err := unstructured.SetNestedField(unstructuredResource.Object, unstructuredResource.GetGeneration(), "status", "observedGeneration"); err != nil {
+				return err
+			}
+
+			// Restore the changes back to the original object.
```

Comment:
Unfortunately we can't patch with the unstructured counterpart of a typed object, because the converted we're using today doesn't fill TypeMeta information.

We've gotten around to this by using ToUnstructured/FromUnstructured. I thought about getting a scheme in but that would be a breaking change for clients.